src/policy/policy.cpp

```diff
@@ -154,6 +154,126 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
+            // Signature must be <= 73 bytes (BIP66). Since signature size is variable, it is not impossible to bloat a
+            // 70-byte signature to 73-byte and trigger fee rejection for marginal cases but the impact is negligible.
+            if (witnessversion == 0 && witnessprogram.size() == 20) {
+                if (tx.wit.vtxinwit[i].scriptWitness.stack.size() != 2)
+                    return TX_BAD_WITNESS;
+                std::vector<unsigned char> hashPubKey(20);
+                std::vector<unsigned char> pubKey = tx.wit.vtxinwit[i].scriptWitness.stack[1];
+                if (tx.wit.vtxinwit[i].scriptWitness.stack[0].size() > 73)
+                    return TX_BAD_WITNESS;
+                CHash160().Write(begin_ptr(pubKey), pubKey.size()).Finalize(begin_ptr(hashPubKey));
+                if (hashPubKey != witnessprogram)
+                    return TX_BAD_WITNESS;
+            }
+
+            // P2WSH requires at least 1 witness stack item but we already know the witness is not empty.
+            // Make sure the witnessScript size is <= 10000 bytes (MAX_SCRIPT_SIZE) and matches witness program.
+            // Max standard witnessScript size is 3600 bytes
+            if (witnessversion == 0 && witnessprogram.size() == 32) {
+                CScript witnessScript = CScript(tx.wit.vtxinwit[i].scriptWitness.stack.back().begin(), tx.wit.vtxinwit[i].scriptWitness.stack.back().end());
+                uint256 hashWitnessScript;
+                if (witnessScript.size() > MAX_SCRIPT_SIZE)
+                    return TX_BAD_WITNESS;
+                if (witnessScript.size() > MAX_STANDARD_P2WSH_SCRIPT_SIZE)
+                    ret = TX_NONSTD_BIG_P2WSH;
+                CSHA256().Write(begin_ptr(witnessScript), witnessScript.size()).Finalize(hashWitnessScript.begin());
+                if (memcmp(hashWitnessScript.begin(), &witnessprogram[0], 32))
+                    return TX_BAD_WITNESS;
+
+                /*
+                 * Count the witness stack size without the witnessScript.
+                 *
+                 * Max standard witness stack size is 100.
+                 * Due to the implicit CLEANSTACK rule and nOpCount limit of 201, the maximum consensus valid
+                 * witness stack size is 604, with 201 OP_CHECKMULTISIGVERIFY in witnessScript.
+                 * Each 0-of-0 OP_CHECKMULTISIGVERIFY is considered as 1 nOpCount, and will remove 3 items from stack:
+                 * dummy item, nSig, and nKey. By putting a true item at the beginning, the script is valid by
+                 * consensus.
+                 * Anything bigger than 604 is guaranteed to leave more than one stack item after execution.
+                 *
+                 * Max consensus size limit for each element is 520 bytes
+                 * Max standard size limit for each element is 80 bytes
+                 */
+                size_t sizeWitnessStack = tx.wit.vtxinwit[i].scriptWitness.stack.size() - 1;
```

Comment:
Following my previous comment, imho, verification of limits like that should be included into EvalScript, as we are already doing for some other limits.